tests/delete/models.py

```diff
@@ -241,3 +241,99 @@ class GenericDeleteBottomParent(models.Model):
     generic_delete_bottom = models.ForeignKey(
         GenericDeleteBottom, on_delete=models.CASCADE
     )
+
```

Comment:
Hi Akash 👋 thank you for being patient, I have started to test this PR.

In this file there is a model called `A` which has every on_delete possible except all of these new ones.
This is actually quite a similar approach to how I am testing in my local project.
It might make sense to review these tests with the model `A` and have each recreated with the DB delete counterparts when this is not already covered by existing tests.

 

For example, I think there might be an issue with the `DB_RESTICT` and `DB_CASCADE` relationship and we can add a test for very similar to `delete.tests.OnDeleteTests.test_restrict_path_cascade_direct`.

Here:

```diff
diff --git a/tests/delete/models.py b/tests/delete/models.py
index 5a3858d7ad..33f8094b6c 100644
--- a/tests/delete/models.py
+++ b/tests/delete/models.py
@@ -58,6 +58,7 @@ class A(models.Model):
         related_name="setnull_nullable_set",
     )
     cascade = models.ForeignKey(R, models.CASCADE, related_name="cascade_set")
+    db_cascade = models.ForeignKey(R, models.DB_CASCADE, related_name="db_cascade_set")
     cascade_nullable = models.ForeignKey(
         R, models.CASCADE, null=True, related_name="cascade_nullable_set"
     )
@@ -67,6 +68,9 @@ class A(models.Model):
     restrict = models.ForeignKey(
         R, models.RESTRICT, null=True, related_name="restrict_set"
     )
+    db_restrict = models.ForeignKey(
+        R, models.DB_RESTRICT, null=True, related_name="db_restrict_set"
+    )
     donothing = models.ForeignKey(
         R, models.DO_NOTHING, null=True, related_name="donothing_set"
     )
@@ -100,9 +104,11 @@ def create_a(name):
         "setdefault",
         "setdefault_none",
         "cascade",
+        "db_cascade",
         "cascade_nullable",
         "protect",
         "restrict",
+        "db_restrict",
         "donothing",
         "o2o_setnull",
     ):
diff --git a/tests/delete/tests.py b/tests/delete/tests.py
index e9a1c09240..bd6d671b2f 100644
--- a/tests/delete/tests.py
+++ b/tests/delete/tests.py
@@ -271,6 +271,26 @@ class OnDeleteTests(TestCase):
         self.assertFalse(A.objects.filter(name="restrict").exists())
         self.assertFalse(R.objects.filter(pk=a.restrict_id).exists())

+    def test_db_restrict_path_db_cascade_direct(self):
+        a = create_a("db_restrict")
+        a.db_restrict.p = P.objects.create()
+        a.db_restrict.save()
+        a.db_cascade_p = a.db_restrict.p
+        a.save()
+        a.db_restrict.p.delete()
+        self.assertFalse(A.objects.filter(name="db_restrict").exists())
+        self.assertFalse(R.objects.filter(pk=a.db_restrict_id).exists())
+
+    def test_db_restrict_path_cascade_direct(self):
+        a = create_a("db_restrict")
+        a.db_restrict.p = P.objects.create()
+        a.db_restrict.save()
+        a.cascade_p = a.db_restrict.p
+        a.save()
+        a.db_restrict.p.delete()
+        self.assertFalse(A.objects.filter(name="db_restrict").exists())
+        self.assertFalse(R.objects.filter(pk=a.db_restrict_id).exists())
+
     def test_restrict_path_cascade_indirect_diamond(self):
         delete_top = DeleteTop.objects.create()
         b1 = B1.objects.create(delete_top=delete_top)
```

The test `test_db_restrict_path_db_cascade_direct` fails with this error on SQLite:
```
ERROR: test_db_restrict_path_db_cascade_direct (delete.tests.OnDeleteTests.test_db_restrict_path_db_cascade_direct)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\boyce\OneDrive\Documents\GitHub\django\django\db\backends\utils.py", line 105, in _execute
    return self.cursor.execute(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\boyce\OneDrive\Documents\GitHub\django\django\db\backends\sqlite3\base.py", line 354, in execute
    return super().execute(query, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sqlite3.IntegrityError: FOREIGN KEY constraint failed

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<path_to_django>\tests\delete\tests.py", line 280, in test_db_restrict_path_cascade_direct
    a.db_restrict.p.delete()
  File "<path_to_django>\django\db\models\base.py", line 1249, in delete
    return collector.delete()
           ^^^^^^^^^^^^^^^^^^
  File "<path_to_django>\django\db\models\deletion.py", line 530, in delete
    count = query.delete_batch(pk_list, self.using)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<path_to_django>\django\db\models\sql\subqueries.py", line 42, in delete_batch
    num_deleted += self.do_query(
                   ^^^^^^^^^^^^^^
  File "<path_to_django>\django\db\models\sql\subqueries.py", line 20, in do_query
    cursor = self.get_compiler(using).execute_sql(CURSOR)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<path_to_django>\django\db\models\sql\compiler.py", line 1570, in execute_sql
    cursor.execute(sql, params)
  File "<path_to_django>\django\db\backends\utils.py", line 79, in execute
    return self._execute_with_wrappers(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<path_to_django>\django\db\backends\utils.py", line 92, in _execute_with_wrappers
    return executor(sql, params, many, context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<path_to_django>\django\db\backends\utils.py", line 100, in _execute
    with self.db.wrap_database_errors:
  File "<path_to_django>\django\db\utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "<path_to_django>\django\db\backends\utils.py", line 105, in _execute
    return self.cursor.execute(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<path_to_django>\django\db\backends\sqlite3\base.py", line 354, in execute
    return super().execute(query, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
django.db.utils.IntegrityError: FOREIGN KEY constraint failed

----------------------------------------------------------------------
```
But `test_db_restrict_path_cascade_direct` passes. I think I would expect the behaviour to be the same and that they would both pass.

Can you confirm this?

Reply:
Thank you for the review @sarahboyce , I think you forgot to add the `db_cascade_p ` field in the model `A`, Updated the code with suggested tests - https://github.com/django/django/pull/16851/files#diff-a842e4474a8c70b274fabec1b44e377a498181dfe6c925801b7dc930917c3e03R324-R343

Reply:
You're completely right thank you ⭐

Reply:
Looked into it again. I still think there's something going on and these tests (might) show it:

```diff
diff --git a/tests/delete/tests.py b/tests/delete/tests.py
index f84831dea7..5b08357f4d 100644
--- a/tests/delete/tests.py
+++ b/tests/delete/tests.py
@@ -331,6 +331,22 @@ class OnDeleteTests(TestCase):
         self.assertFalse(A.objects.filter(name="db_restrict").exists())
         self.assertFalse(R.objects.filter(pk=a.db_restrict_id).exists())

+    def test_db_restrict_db_cascade_direct(self):
+        a = create_a("db_restrict")
+        a.db_cascade = a.db_restrict
+        a.save()
+        a.db_restrict.delete()
+        self.assertFalse(A.objects.filter(name="db_restrict").exists())
+        self.assertFalse(R.objects.filter(pk=a.db_restrict_id).exists())
+
+    def test_db_restrict_cascade_direct(self):
+        a = create_a("db_restrict")
+        a.cascade = a.db_restrict
+        a.save()
+        a.db_restrict.delete()
+        self.assertFalse(A.objects.filter(name="db_restrict").exists())
+        self.assertFalse(R.objects.filter(pk=a.db_restrict_id).exists())
+
     def test_db_restrict_path_cascade_direct(self):
         a = create_a("db_restrict")
         a.db_restrict.p = P.objects.create()
```
I have the FK error with the `test_db_restrict_db_cascade_direct` test but not `test_db_restrict_cascade_direct`

In my local project I have:
`models.py`
```python
from django.db import models


class Basic(models.Model):
    name = models.CharField(max_length=30)


class EveryOnDelete(models.Model):
    fk_cascade = models.ForeignKey(
        Basic,
        on_delete=models.CASCADE,
        null=True,
        related_name="fk_cascade",
    )
    fk_db_cascade = models.ForeignKey(
        Basic,
        on_delete=models.DB_CASCADE,
        null=True,
        related_name="fk_db_cascade",
    )
    fk_set_null = models.ForeignKey(
        Basic,
        on_delete=models.SET_NULL,
        null=True,
        related_name="fk_set_null",
    )
    fk_db_set_null = models.ForeignKey(
        Basic,
        on_delete=models.DB_SET_NULL,
        null=True,
        related_name="fk_db_set_null",
    )
    fk_set_default = models.ForeignKey(
        Basic,
        on_delete=models.SET_DEFAULT,
        default="1",
        null=True,
        related_name="fk_set_default",
    )
    fk_db_set_default = models.ForeignKey(
        Basic,
        on_delete=models.DB_SET_DEFAULT,
        null=True,
        db_default="1",
        related_name="fk_db_set_default",
    )
    fk_restrict = models.ForeignKey(
        Basic,
        on_delete=models.RESTRICT,
        null=True,
        related_name="fk_restrict",
    )
    fk_db_restrict = models.ForeignKey(
        Basic,
        on_delete=models.DB_RESTRICT,
        null=True,
        related_name="fk_db_restrict",
    )
    fk_protect = models.ForeignKey(
        Basic,
        on_delete=models.PROTECT,
        null=True,
        related_name="fk_protect",
    )
```
and some tests
```python
from django.test import TestCase
from .models import Basic, EveryOnDelete


class DBDeleteTests(TestCase):
    @classmethod
    def setUpTestData(cls):
        Basic.objects.create(name="id-1", id="1")

# Removed some to make this smaller

    def test_db_fk_restrict_basic_deleted_allowed_cascade(self):
        basic_1 = Basic.objects.create(name="not-protected")
        EveryOnDelete.objects.create(
            fk_db_restrict=basic_1,
            fk_cascade=basic_1,
        )
        basic_1.delete()
        self.assertEqual(Basic.objects.count(), 1)
        self.assertEqual(EveryOnDelete.objects.count(), 0)

    def test_db_fk_restrict_basic_deleted_allowed_db_cascade(self):
        basic_1 = Basic.objects.create(name="not-protected")
        EveryOnDelete.objects.create(
            fk_db_cascade=basic_1,
            fk_db_restrict=basic_1,
        )
        basic_1.delete()
        self.assertEqual(Basic.objects.count(), 1)
        self.assertEqual(EveryOnDelete.objects.count(), 0)
```
Here `test_db_fk_restrict_basic_deleted_allowed_db_cascade` is failing with the FK error, I think it should pass rather than raise an `IntegrityError`?

Reply:
Got some time this week after a long rush, will look into this. Thank you for finding it out @sarahboyce

Reply:
I have looked into this issue, the current implementation seems to be like this - 

There are four scenarios regarding this:
1. The ForeignKey parent has Non DB On Delete and the Child has Non DB On Delete - Allow
2. The ForeignKey parent has DB level On Delete and the Child has DB level On Delete - Allow 
3. The ForeignKey parent has DB level On Delete and the Child has Non DB On Delete - Not Allowed
(The check should throw an error, as if the Child is deleted by internal database operations, it might lead to
Integrity Error in those tables which references the child itself)
4. The ForeignKey parent has Non DB On Delete and the Child has DB level On Delete - Allow

Now the problem arises in only in the fourth scenario, as the 3rd scenario is not allowed and the first two comes up with a very consistent approach.

Let us assume there are three models
```python
class Parent(models.Model):
    pass

class Child(models.Model):
   parent = models.ForeignKey(Parent, models.CASCADE, null=True)

class GrandChild(models.Model):
   child = models.ForeignKey(Child, models.DB_RESTRICT, null=True)
```

As the django Non DB cascade will internally delete all the related objects, it will not care about the DB Restrict just like the internal RESTRICT, so I have to cover the scenario of `DB_RESTRICT` alone and not allow deletion if any of the related objects have `DB_RESTRICT` as FK